provider/oci/instance.go

```diff
@@ -4,43 +4,295 @@
 package oci
 
 import (
+	"context"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
 	"github.com/juju/errors"
+	"github.com/juju/juju/status"
 
+	"github.com/juju/juju/environs/instances"
 	"github.com/juju/juju/instance"
 	"github.com/juju/juju/network"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
+)
+
+const (
+	// MinVolumeSizeMB is the minimum size in MB for a volume or boot disk
+	MinVolumeSizeMB = 51200
+
+	// MaxVolumeSizeMB is the maximum size in MB for a volume or boot disk
+	MaxVolumeSizeMB = 16777216
 )
 
-type ociInstance struct{}
+type ociInstance struct {
+	arch     *string
+	instType *instances.InstanceType
+	env      *Environ
+	mutex    *sync.Mutex
+	ocid     string
+	etag     *string
+	raw      ociCore.Instance
+}
+
+type vnicWithIndex struct {
+	Vnic ociCore.Vnic
+	Idx  int
+}
 
 var _ instance.Instance = (*ociInstance)(nil)
-var _ instance.InstanceFirewaller = (*ociInstance)(nil)
+
+// newInstance returns a new oracleInstance
+func newInstance(raw ociCore.Instance, env *Environ) (*ociInstance, error) {
+	if raw.Id == nil {
+		return nil, errors.New(
+			"Instance response does not contain an ID",
+		)
+	}
+	mutex := &sync.Mutex{}
+	arch := "amd64"
+	instance := &ociInstance{
+		raw:   raw,
+		mutex: mutex,
+		env:   env,
+		arch:  &arch,
+		ocid:  *raw.Id,
+	}
+
+	return instance, nil
+}
+
+// SetInstance sets the raw property of ociInstance{}
+// Testing purposes.
+func (o *ociInstance) SetInstance(inst ociCore.Instance) {
+	o.raw = inst
+}
+
+func (o *ociInstance) availabilityZone() string {
+	return *o.raw.AvailabilityDomain
+}
 
 // Id implements instance.Instance
 func (o *ociInstance) Id() instance.Id {
-	return instance.Id(0)
+	if o.raw.Id == nil {
+		return instance.UnknownId
+	}
+	return instance.Id(*o.raw.Id)
 }
 
 // Status implements instance.Instance
 func (o *ociInstance) Status() instance.InstanceStatus {
-	return instance.InstanceStatus{}
+	// This should not happen, unless someone bypassed newInstance()
+	// and created the ociInstance{} object manually.
+	if o.raw.Id == nil {
+		if err := o.refresh(); err != nil {
```

Comment:
refresh used to use ```o.ocid```. Even if someone composed ```ociInstance{}``` manually, without any value in ```o.raw```, this would have attempted to populate ```o.raw``` using ```o.ocid``` to do an API call. Was trying to be forgiving here, but you inspired me :). Removed ```o.ocid```. All instances of ```ociInstance{}``` must be created through ```newInstance()```.